yt_dlp/extractor/neteasemusic.py

```diff
@@ -66,45 +66,40 @@ def _download_eapi_json(self, path, video_id, query_body, headers={}, **kwargs):
                 **headers,
             }, **kwargs)
 
-    def _call_player_api(self, song_id, bitrate):
+    def _call_player_api(self, song_id, level):
         return self._download_eapi_json(
-            '/song/enhance/player/url', song_id, {'ids': f'[{song_id}]', 'br': bitrate},
-            note=f'Downloading song URL info: bitrate {bitrate}')
+            '/song/enhance/player/url/v1', song_id, {'ids': f'[{song_id}]', 'level': level, 'encodeType': 'flac'},
+            note=f'Downloading song URL info: level {level}')
 
-    def extract_formats(self, info):
-        err = 0
+    def _extract_formats(self, info):
         formats = []
+        format_MD5s = []
         song_id = info['id']
-        for song_format in self._FORMATS:
-            details = info.get(song_format)
-            if not details:
+        for level in self._LEVELS:
+            song = traverse_obj(self._call_player_api(song_id, level), ('data', 0, {dict})) or {}
+            if song.get('level') in [song_format.get('level') for song_format in formats]:
+                break  # API returns existing level - the user have no access to further ones
+            if not url_or_none(song.get('url')):
                 continue
-            bitrate = int_or_none(details.get('bitrate')) or 999000
-            for song in traverse_obj(self._call_player_api(song_id, bitrate), ('data', lambda _, v: url_or_none(v['url']))):
-                song_url = song['url']
-                if self._is_valid_url(song_url, info['id'], 'song'):
-                    formats.append({
-                        'url': song_url,
-                        'format_id': song_format,
-                        'asr': traverse_obj(details, ('sr', {int_or_none})),
-                        **traverse_obj(song, {
-                            'ext': ('type', {str}),
-                            'abr': ('br', {self.kilo_or_none}),
-                            'filesize': ('size', {int_or_none}),
-                        }),
-                    })
-                elif err == 0:
-                    err = traverse_obj(song, ('code', {int})) or 0
-
+            if 'md5' in song and song['md5'] in format_MD5s:
+                continue  # API returns existing audio file - this level is the same as existing ones
```

Comment:
if we already `break` upon reaching a dupe, when will this code path be reached?

Reply:
the idea with breaking the loop is so that we don't make all 7 api calls unnecessarily

Reply:
Yeah I'm probably not in the best mind when reasoning the logic while writing this... sorry. Reverted.

Reply:
the company was annoying, and he didn't have an interface that could directly display all the sound quality options.
The official client also uses traversal.I just went to decompile it